moosez/predict.py

```diff
@@ -69,87 +70,81 @@ def predict(model_name: str, input_dir: str, output_dir: str, accelerator: str)
-
-    resampled_image = ImageResampler.resample_image(moose_img_object=moose_image_object,
-                                                    interpolation=constants.INTERPOLATION,
-                                                    desired_spacing=desired_spacing)
-    # if model name has body in it, run logic below
-    if "body" in model_name:
-        image_processing.write_image(resampled_image, os.path.join(temp_folder, constants.RESAMPLED_IMAGE_FILE_NAME),
-                                     False, False)
-    else:
-        image_processing.write_image(resampled_image, os.path.join(temp_folder, constants.RESAMPLED_IMAGE_FILE_NAME),
-                                     moose_image_object.is_large, False)
-    return temp_folder, resampled_image, moose_image_object
-
-
-def postprocess(original_image: str, output_dir: str, model_name: str) -> None:
-    """
-    Postprocesses the predicted images.
-
-    :param original_image: The path to the original image.
-    :type original_image: str
-    :param output_dir: The output directory containing the label image.
-    :type output_dir: str
-    :param model_name: The name of the model.
-    :type model_name: str
-    :return: None
-    :rtype: None
-    """
-    # [1] Resample the predicted image to the original image's voxel spacing
-    predicted_image = file_utilities.get_files(output_dir, '.nii.gz')[0]
-    original_header = nib.load(original_image).header
-    native_spacing = original_header.get_zooms()
-    native_size = original_header.get_data_shape()
-    resampled_prediction = ImageResampler.resample_segmentations(input_image_path=predicted_image,
-                                                                 desired_spacing=native_spacing,
-                                                                 desired_size=native_size)
-    multilabel_image = os.path.join(output_dir, MODELS[model_name]["multilabel_prefix"] +
-                                    os.path.basename(original_image))
-    # if model_name has body in it, run logic below
-    if "body" in model_name:
-        resampled_prediction_data = resampled_prediction.get_fdata()
-        resampled_prediction_new = nib.Nifti1Image(resampled_prediction_data, resampled_prediction.affine,
-                                                   resampled_prediction.header)
-        image_processing.write_image(resampled_prediction_new, multilabel_image, False, True)
-    else:
-        image_processing.write_image(resampled_prediction, multilabel_image, False, True)
-    os.remove(predicted_image)
-
-
-def count_output_files(output_dir: str) -> int:
+    original_image_files = file_utilities.get_files(original_image_directory, ".nii.gz")
+
+    resample_transform = ImageResampler.monai_resampling(
+        interpolation=constants.INTERPOLATION,
+        desired_spacing=voxel_spacing,
+        device=accelerator,
+        output_dir=temp_folder)
+    resampled_image = resample_transform(original_image_files[0])
```

Comment:
Hi @ltetrel the code seems to get stuck here for me!

@mprires @Keyn34 can you confirm? does this happen to you as well?

Reply:
Same problem on my end. Testing on Ubuntu 22.04, 256 GB of RAM and Nvidia 3090. 
Screenshot of adapted code to check for the line where it would not get past:
<img width="1135" alt="Screenshot 2024-04-12 at 17 33 00" src="https://github.com/ENHANCE-PET/MOOSE/assets/48754309/d1f24e3f-d5d4-4346-b279-ea19fca660c0">

Screenshot of the terminal after it stopped proceeding:
<img width="971" alt="Screenshot 2024-04-12 at 17 34 27" src="https://github.com/ENHANCE-PET/MOOSE/assets/48754309/89ec6a28-0982-49ea-80d1-62382769fa8c">

Also had to wait quite a bit for copy to install

Reply:
For cupy, I am not sure why I needed this. I will double check where it is being used (I suppose monai processing but not sure).

Regarding the current issue:
1. Were you able to run and pass the tests ?
2. I used `linear` resampling in my case because of VRAM limitations, currently in the process to test on a better GPU. Can you also try with `linear` ?

Reply:
@ltetrel 
1. It only got through the first print, it was stuck there after
2. I will try linear resampling.


@LalithShiyam I haven't tested on a windows machine. I will test it later today and update this thread

Reply:
I will try again from scratch when I have access to the GPU server.
**cupy:**
For the moment I tested with `Ubuntu 20.04.6 LTS`. While I am thinking of it, I was still using my "old" environment (from previous moose version) with python3.9, I should try with 3.10. @LalithShiyam why upgrading to 3.10 ? Can you also try with 3.9 just to make sure ?

Reply:
**stuck at resampling**:
Maybe the culprit is the parallelization of the patch saving process. It reminds me some issues I had at some point where the thread would not sent its finished status for whatever reason (maybe some conflicts with monai internal threading management)... If you want to quickly bypass the multiprocessing, hardcode `self.num_workers = 1`. I will work on this issue when I have hardware.
https://github.com/ENHANCE-PET/MOOSE/blob/f9ac1dfeadfc6b28c283820d9d882f41ff0afb2e/moosez/image_processing.py#L972-L977

Another solution is to use monai internal multithreading mechanism through dataset loaders, this is ideal but complicated to use with the current loop-based implementation of moose.

Ultimately, we can remove the image saving process (and multithreading) because we can bypass the disk caching entirely by calling nnunet directly and pass numpy arrays.

Reply:
@LalithShiyam @mprires 
cupy installation should be much faster with latest commit
https://github.com/ENHANCE-PET/MOOSE/pull/115/commits/3f91216ec153e9d4f6bf003166dcc532e621a689

I was not able to re-produce your issues, I am runing on a fresh VM with ubuntu22, python3.10, cuda_11.7.
Can you re-try with latest commit, a new venv and give me more information on your system ?
![image](https://github.com/ENHANCE-PET/MOOSE/assets/37963074/b0ac64fe-83c8-4356-9262-e064181fbedd)

Reply:
Will do so tomorrow. Today doesn't seem to end.

Reply:
Just updated to cuda 12.4, and it is working on my end on a fresh install. Can you re-try ?

Reply:
Absolutely. @mprires do we have space in mz or should we use kylo?

Reply:
I will test on kylo now and let you know

Reply:
I need to update cuda on kylo (we are on 12.2 both on kylo and mz). I tried on cuda 12.2 on mz. Instalation is a breeze now. On CUDA 12.2 I do get the this error. Will try on CUDA 12.4 ASAP and update.
![Screenshot 2024-04-30 at 18 29 03](https://github.com/ENHANCE-PET/MOOSE/assets/48754309/8849f2b9-8e2f-42f1-9b8b-2e4d7e97065b)
![Screenshot 2024-04-30 at 18 29 29](https://github.com/ENHANCE-PET/MOOSE/assets/48754309/285af16a-0ef5-421e-830b-f19841e030bb)

Reply:
Can you retry now that preproc multithreading is disabled ? I can add an issue related to the current deadlocks issues in python multiprocess if needed.
@Keyn34 @mprires